ASoC: SOF: Intel: HDA: Set IPC4-specific DSP ops for CaVS platforms

ujfalusi opened the pull request:
Add implementation of low level, platform dependent IPC4 message handling
and set the DSP ops for IPC4 for APL, CNL and TGL platforms.

Co-developed-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>
Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>
Signed-off-by: Ranjani Sridharan <ranjani.sridharan@linux.intel.com>

bardliao commented on sound/soc/sof/intel/cnl.c:
```diff
@@ -29,6 +30,65 @@ static const struct snd_sof_debugfs_map cnl_dsp_debugfs[] = {
+			sdev->ipc->msg.rx_data = NULL;
+		}
+
+		cnl_ipc_host_done(sdev);
```
Just curious, do we need spin_lock_irq(&sdev->ipc_lock); for cnl_ipc_host_done(sdev);?

ujfalusi replied:
We were not doing that in IPC3, do you think we should in IPC4?

bardliao replied:
> We were not doing that in IPC3, do you think we should in IPC4?

No, my question was for both IPC3 and IPC4. I don't know if it is possible that something can interleave when we set the busy and done bits.

ujfalusi replied:
@bardliao, What I know that we can receive notification between sending a message and getting back the reply. I believe we have the higher level serialization.
we must not block the rx, which is async in many ways.

So far it worked fine, but we can take a deeper look and address both IPCs if there is a need?

bardliao replied:
> @bardliao, What I know that we can receive notification between sending a message and getting back the reply. I believe we have the higher level serialization. we must not block the rx, which is async in many ways.
> 
> So far it worked fine, but we can take a deeper look and address both IPCs if there is a need?

Sure

ujfalusi replied:
@bardliao, I think the IPC4 logic is not correct. I'll update the PR when I figured it out, but in short:
`(hipcida & CNL_DSP_REG_HIPCIDA_DONE)` indicates that the DSP acked a tx (received the message), nothing else.
`(hipctdr & CNL_DSP_REG_HIPCTDR_BUSY)` indicates that the DSP send a reply _or_ a new notification.

ranj063 replied:
good catch @ujfalusi

RanderWang replied:
that is why I was first confused

RanderWang commented on sound/soc/sof/intel/cnl.c:
```diff
@@ -29,6 +30,65 @@ static const struct snd_sof_debugfs_map cnl_dsp_debugfs[] = {
+			spin_lock_irq(&sdev->ipc_lock);
+
+			snd_sof_ipc_get_reply(sdev);
+			snd_sof_ipc_reply(sdev, data->primary);
```
where do you process reply status ?  It is not correct to process reply message here.  FW send another ipc message for the status of last message. It should be process in  else path, not here.  I meant rx message path

RanderWang replied:
you can send a error ipc message to validate this PR

RanderWang replied:
ok, I understand your code

ujfalusi replied:
@RanderWang, during audio playback I can use the msg inject.
Sending PAUSE for the pipeline (0x13010003) - supported, success:
```
[  132.806420] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x13010003|0x0: GLB_SET_PIPELINE_STATE
[  132.807194] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x33000000|0x0: GLB_SET_PIPELINE_STATE
[  132.807211] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx done : 0x13010003|0x0: GLB_SET_PIPELINE_STATE
```

Sending RESET to the pipeline (0x13010002) - not supported, fail:
```
[  655.677763] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx      : 0x13010002|0x0: GLB_SET_PIPELINE_STATE
[  655.678219] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc tx reply: 0x33000006|0x1: GLB_SET_PIPELINE_STATE
[  655.678226] sof-audio-pci-intel-tgl 0000:00:1f.3: FW reported error: 6 - Unknown error while processing the request
[  655.678241] sof-audio-pci-intel-tgl 0000:00:1f.3: ipc error for msg 0x13010002|0x0
[  655.678246] snd_sof_ipc_msg_injector snd_sof.msg_injector.0: IPC message send failed: -22
```
Do we need to read something else after this?

ranj063 commented:
SOFCI TEST

ujfalusi commented:
Changes since v1:
- Rewrite the interrupt handlers for correct handling of the message reception

When IPC4 is used, the interrupt reasons are used a bit differently. One source is just telling the host that the DSP received the message and started to process it (CNL_DSP_REG_HIPCIDA_DONE or HDA_DSP_REG_HIPCIE_DONE), the other tells that the DSP sent a message which can be a reply or a notification.

plbossart reviewed (approved):
approving blindly.
